internal/service/cluster/cluster.go

```diff
@@ -210,23 +217,40 @@ func (s *Service) Run(ctx context.Context, host service.Host) error {
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
+	limiter := rate.NewLimiter(rate.Every(stateUpdateMinInterval), 1)
 	s.node.Observe(ckit.FuncObserver(func(peers []peer.Peer) (reregister bool) {
+		tracer := s.tracer.Tracer("")
+		spanCtx, span := tracer.Start(ctx, "NotifyClusterChange", trace.WithSpanKind(trace.SpanKindInternal))
+		defer span.End()
+
+		if s.opts.EnableStateUpdatesLimiter {
+			// Limit how often we notify components about peer changes. At start up we may receive N updates in a period
+			// of less than one second. This leads to a lot of unnecessary processing.
+			_, span := tracer.Start(spanCtx, "RateLimitWait", trace.WithSpanKind(trace.SpanKindInternal))
+			if err := limiter.Wait(ctx); err != nil {
+				// This should never happen, but it should be safe to just ignore it and continue.
+				level.Warn(s.log).Log("msg", "failed to wait for rate limiter on peers update", "err", err)
+				span.RecordError(err)
+			}
+			span.End()
+
+			// Grab a fresh view of the peers for logging, since we may have waited a bit for the limiter to permit.
+			peers = s.node.Peers()
```

Comment:
Hm, this won't cause any bugs but it's really not the intended way to use Observe :thinking: If there are new peers, Observe will be called again and it'll catch up over time. 

I'm fine with this though.

Reply:
Yeah, if we're fine with the logged peers being potentially not the most up-to-date, I'd be fine with getting rid of this. It may just be confusing for the next person debugging this.

Reply:
I'd be cool with either keeping this in or replacing it with a comment explaining what I just said (that there may have been new events in the last second and Observe will be called again to catch up)

Reply:
Sounds good, after thinking about this I opted to not re-lookup the peers and add a comment.